### PR TITLE
Pre-scale attention query to prevent fp16 overflow

### PR DESCRIPTION
**Pull request by fcakyon**
Pre-scale the attention query before the QK^T matmul in `Attention` and `AAttn` so the raw dot product cannot exceed the fp16 max (65504) and overflow to inf/nan under AMP. We saw the unscaled QK^T overflow on a fraction of mixed-precision batches and turn the softmax into NaN, which forced an fp32 fallback. The reorder is exact-identical in fp32 (verified max-abs-diff 0.0, so model weights and outputs are unaffected) and keeps the product in range under mixed precision.

Pre-scaling the query is the standard scaled-attention formulation, for example Apple MobileCLIP [`mobileclip/models/mci.py`](https://github.com/apple/ml-mobileclip/blob/main/mobileclip/models/mci.py#L141): `attn = (q * self.scale) @ k.transpose(-2, -1)`.

```python
# before
attn = (q.transpose(-2, -1) @ k) * self.scale             # Attention
attn = (q.transpose(-2, -1) @ k) * (self.head_dim**-0.5)   # AAttn

# after
attn = (q * self.scale).transpose(-2, -1) @ k             # Attention
attn = (q * (self.head_dim**-0.5)).transpose(-2, -1) @ k   # AAttn
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
⚡ This PR makes a small but important optimization to attention computation in `block.py` by applying the scaling factor earlier in the matrix multiplication pipeline.

### 📊 Key Changes
- Updated two attention `forward()` methods in `ultralytics/nn/modules/block.py`.
- Changed attention score computation from:
  - scaling the result after `q @ k`
- To:
  - scaling `q` first, then performing the matrix multiplication
- Specifically, these lines were refactored:
  - `attn = (q.transpose(-2, -1) @ k) * self.scale` → `attn = (q * self.scale).transpose(-2, -1) @ k`
  - `attn = (q.transpose(-2, -1) @ k) * (self.head_dim**-0.5)` → `attn = (q * (self.head_dim**-0.5)).transpose(-2, -1) @ k`
- No model architecture, APIs, or outputs are intentionally changed 🔧

### 🎯 Purpose & Impact
- Improves implementation efficiency by moving the scaling step before matrix multiplication 🚀
- Keeps the attention logic mathematically equivalent, so behavior should remain the same for users
- May help low-level performance, numerical handling, or backend optimization depending on hardware and PyTorch execution
- Safe internal cleanup: no user-facing workflow changes, retraining steps, or config updates expected ✅

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @fcakyon, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help streamline review; an engineer will assist you soon. Please review the checklist below 😊
- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Looks good. This is a small, targeted numerical-stability change, and the pre-scaling is algebraically equivalent to the previous post-matmul scaling while reducing fp16 overflow risk in the attention score computation. I don't see any actionable issues in the diff.

**Review by glenn-jocher — APPROVED** [review]
LGTM. This is a minimal algebraic reorder at the source of the fp16 overflow: scale before the attention matmul rather than after the potentially overflowing product. I verified the representative half-precision case where post-scale overflows to  while pre-scale remains finite.

**Review by glenn-jocher — APPROVED** [review]
LGTM. This is a minimal algebraic reorder at the source of the fp16 overflow: scale before the attention matmul rather than after the potentially overflowing product. I verified the representative half-precision case where post-scale overflows to inf while pre-scale remains finite.

**Comment by UltralyticsAssistant** [comment]
Merged! 🎉 Thanks @fcakyon for this clean attention-path optimization in `block.py`, and thanks @glenn-jocher for the collaboration and review.

As Antoine de Saint-Exupéry said, *“Perfection is achieved, not when there is nothing more to add, but when there is nothing left to take away.”* This PR is a great example of that idea in practice: a small internal refinement that keeps behavior unchanged while improving implementation efficiency and maintaining the elegance of the codebase.

Really appreciate the thoughtful contribution and continued focus on making Ultralytics better, one sharp optimization at a time 🚀
